### PR TITLE
mc-mint-auditor: add useful traits, _hex suffix, MintTxNonce, some cleanups

### PR DESCRIPTION
**Pull request by eranrund**
### Motivation

This PR contains a bunch of changes that I accumulated in my branch that does auditing of MintTxs<->GnosisSafeDeposits. Since it was getting large I figured I will start by getting the unrelated changes in.

The stuff here includes:
- Adding a bunch of useful trait derives on the db models (e.g. Clone, Hash, etc)
- Creating custom Diesel types for wrapping `EthAddr` and `EthTxHash`. This saves us from a bunch of conversions all over the place.
- Standardizing on `_addr`. Previously, there was a mix of `_addr` and `_address`. `_addr` is a well-established abbreviation and the names are already long enough so there's no need for extra verbosity everywhere.
- Changing some `insert()` implementations to return the id of the newly inserted objects by storing it in the instance. This comes in handy when we want to reference the object in a SQL statement without figuring out how to re-fetch it.
- Some tests cleanups
- `_hex` suffix for hex-encoded SQL fields
- A new type for representing a `MintTx` nonce that holds information we care about such as the matching Ethereum transaction hash
- Some utilities/error variants that come in handy in followup PRs

### CODE REVIEW

File: mint-auditor/src/gnosis/eth_data_types.rs [hunk 1]
```diff
@@ -58,18 +101,29 @@ impl FromStr for EthTxHash {
...
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl FromRandom for EthTxHash {
+    fn from_random<T: RngCore + CryptoRng>(rng: &mut T) -> Self {
+        let mut result = Self([0u8; Self::LEN]);
+        rng.fill_bytes(&mut result.0);
+        result
```
Comment (remoun) [inline mint-auditor/src/gnosis/eth_data_types.rs:126]: Optional/Equivalent: Use the existing impl for byte arrays.
```suggestion
        Self(FromRandom::from_random(rng))
```

File: mint-auditor/src/gnosis/eth_data_types.rs [hunk 2]
```diff
@@ -3,16 +3,26 @@
...
+    hash::{Hash, Hasher},
+    str::FromStr,
+};
 
 /// Ethereum 20 byte address.
 /// We currently do not store the decoded bytes since we want to maintain the
 /// original capitalization (which is how Ethereum addresses represent a
 /// checksum). We don't have a need for the raw bytes.
-#[derive(Clone, Debug, Default, DeserializeFromStr, SerializeDisplay)]
+#[derive(Clone, Debug, Default, DeserializeFromStr, Ord, PartialOrd, SerializeDisplay)]
```
Comment (remoun) [inline mint-auditor/src/gnosis/eth_data_types.rs:18]: IIUC we'll need to implement `[Partial]Ord` with lowercasing, similar to `[Partial]Eq`. 

The [`PartialOrd` docs](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) state:
> The methods of this trait must be consistent with each other and with those of [`PartialEq`](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html). The following conditions must hold:
> `a == b` if and only if `partial_cmp(a, b) == Some(Equal)`.

File: mint-auditor/src/mint_tx_nonce.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::{gnosis::EthTxHash, Error};
+use mc_transaction_core::mint::constants::NONCE_LENGTH;
+
+/// Data structure for representing what is encoded in a MintTx nonce.
+/// See https://ethereum.org/en/developers/docs/networks/ for description of each network.
```
Comment (remoun) [inline mint-auditor/src/mint_tx_nonce.rs:7]: This line seems out of place
```suggestion
```

File: mint-auditor/src/db/models/gnosis_safe_tx.rs [hunk 4]
```diff
@@ -34,4 +36,11 @@ impl GnosisSafeTx {
 
         Ok(())
     }
+
+    /// Decode a Gnosis Safe transaction.
+    pub fn decode(&self) -> Result<RawGnosisTransaction, Error> {
+        Ok(RawGnosisTransaction::from_json_bytes(
+            self.raw_tx_json.as_bytes(),
```
Comment (remoun) [inline mint-auditor/src/db/models/gnosis_safe_tx.rs:43]: How about we make this operate on `&str` rather than `&[u8]`?

**Review by remoun — COMMENTED** [review]
LGTM overall, just some questions and suggestions.

Nit: This is using `_hex` suffix, not prefix.

File: mint-auditor/src/db/sql_types.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+impl fmt::Display for SqlEthAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl<DB: Backend> ToSql<diesel::sql_types::Text, DB> for SqlEthAddr {
+    fn to_sql<W: Write>(&self, out: &mut Output<W, DB>) -> serialize::Result {
+        <String as ToSql<diesel::sql_types::Text, DB>>::to_sql(&self.0.to_string(), out)
```
[the remark this answers is not in the document]
Comment (eranrund) [reply mint-auditor/src/db/sql_types.rs:59]: No. This was taken from Diesel's examples.

```
59 |         String::to_sql(&self.0.to_string(), out)
   |         ^^^^^^^^^^^^^^ cannot infer type for type parameter `A` declared on the trait `ToSql`
```

```
  --> mint-auditor/src/db/models/sql_types.rs:59:28
    |
59  |         self.0.to_string().to_sql::<diesel::sql_types::Text, DB>(out)
    |                            ^^^^^^                            -- help: remove this generic argument
    |                            |
    |                            expected 1 generic argument
    |
note: associated function defined here, with 1 generic parameter: `W`
```

```
   --> mint-auditor/src/db/models/sql_types.rs:59:62
    |
59  |         self.0.to_string().to_sql::<diesel::sql_types::Text>(out)
    |                            ------                            ^^^ the trait `std::io::Write` is not implemented for `diesel::sql_types::Text`
    |                            |
    |                            required by a bound introduced by this call
    |
note: required by a bound in `diesel::serialize::ToSql::to_sql`
   --> /home/eran/.cargo/git/checkouts/diesel-2c7eb767b8a3657d/026f637/diesel/src/serialize.rs:191:18
    |
191 |     fn to_sql<W: Write>(&self, out: &mut Output<W, DB>) -> Result;
    |                  ^^^^^ required by this bound in `diesel::serialize::ToSql::to_sql`
```

File: mint-auditor/src/gnosis/eth_data_types.rs (hunk 2 above)
Comment (eranrund) [reply mint-auditor/src/gnosis/eth_data_types.rs:18]: Good catch

**Review by remoun — APPROVED** [review]

**Review by garbageslam — APPROVED** [review]

**Pull request merged**